### PR TITLE
Add ability to move features from one data set to another in memory

### PR DESCRIPTION
**Pull request by remcob-gr**
This is useful if features have been previously written out to different binary files.
A usage example is:
```
d1 = Dataset('d1.bin')
d2 = Dataset('d2.bin')
d1.add_features_from(d2)
```
The PR includes the functionality itself (add_features_from in the Python package and the underlying C API call LGBM_DatasetAddFeaturesFrom), as well as unit tests for it.

### CODE REVIEW

File: src/io/dataset.cpp [hunk 1]
```diff
@@ -881,4 +943,68 @@ void Dataset::FixHistogram(int feature_idx, double sum_gradient, double sum_hess
   }
 }
 
+template<typename T>
+void PushVector(std::vector<T>& dest, const std::vector<T>& src){
+  for(auto i : src){
+    dest.push_back(i);
```
Comment (guolinke) [inline src/io/dataset.cpp]: add a `reserve` before push_back?

File: src/io/dataset.cpp [hunk 2]
```diff
@@ -881,4 +943,68 @@ void Dataset::FixHistogram(int feature_idx, double sum_gradient, double sum_hess
...
+
+void Dataset::addFeaturesFrom(Dataset* other){
+  if(other->num_data_ != num_data_){
+    throw std::runtime_error("Cannot add features from other dataset with a different number of rows");
+  }
+  PushVector(feature_names_, other->feature_names_);
+  PushVector(feature2subfeature_, other->feature2subfeature_);
+  PushVector(group_feature_cnt_, other->group_feature_cnt_);
+  for(auto& fg : other->feature_groups_){
+    feature_groups_.push_back(std::move(fg));
```
Comment (guolinke) [inline src/io/dataset.cpp]: move or copy?

File: src/io/dataset.cpp [hunk 3]
```diff
@@ -690,6 +707,51 @@ void Dataset::SaveBinaryFile(const char* bin_filename) {
...
+    fprintf(file, "%s, ", n.c_str());
+  }
+  std::vector<std::unique_ptr<BinIterator>> iterators(num_total_features_);
+  for(int j = 0; j < num_total_features_; j++){
+    if(used_feature_map_[j] >= 0){
+	auto real_feature_idx = used_feature_map_[j];
+	auto group_idx = feature2group_[real_feature_idx];
+	auto sub_idx = feature2subfeature_[real_feature_idx];
+	std::unique_ptr<BinIterator> sub_iter(feature_groups_[group_idx]->SubFeatureIterator(sub_idx));
+	iterators[j] = std::move(sub_iter);
```
Comment (guolinke) [inline src/io/dataset.cpp]: move or copy?

File: src/c_api.cpp [hunk 4]
```diff
@@ -878,6 +889,16 @@ int LGBM_DatasetGetNumFeature(DatasetHandle handle,
   API_END();
 }
 
+int LGBM_DatasetAddFeaturesFrom(DatasetHandle target,
+				 DatasetHandle source) {
+  API_BEGIN();
+  auto target_d = reinterpret_cast<Dataset*>(target);
+  auto source_d = reinterpret_cast<Dataset*>(source);
+  target_d->addFeaturesFrom(source_d);
+  LGBM_DatasetFree(source_d);
```
[the remark this answers is not in the document]
Comment (remcob-gr) [reply src/c_api.cpp]: This was in an attempt to try to conserve memory, but it's probably a premature optimisation.
I'll change it (and change the moves to copies).
Thanks for the review.

File: src/io/dataset.cpp (hunk 3 above)
Comment (remcob-gr) [reply src/io/dataset.cpp]: Move seems OK here, the iterator is allocated on line 738, we want to move ownership of it to the iterators vector.
@guolinke, do you agree?

File: src/io/dataset.cpp (hunk 2 above)
Comment (remcob-gr) [reply src/io/dataset.cpp]: This should be a copy, will fix.

File: src/io/dataset.cpp (hunk 1 above)
Comment (remcob-gr) [reply src/io/dataset.cpp]: Done.

File: src/io/dataset.cpp [hunk 5]
```diff
@@ -690,6 +707,49 @@ void Dataset::SaveBinaryFile(const char* bin_filename) {
...
+  auto file = fopen(text_filename, "wt");
+  fprintf(file, "num_features: %d\n", num_features_);
+  fprintf(file, "num_total_features: %d\n", num_total_features_);
+  fprintf(file, "num_groups: %d\n", num_groups_);
+  fprintf(file, "num_data: %d\n", num_data_);
+  fprintf(file, "feature_names: ");
+  for(auto n : feature_names_){
+    fprintf(file, "%s, ", n.c_str());
+  }
+  fprintf(file, "\nmonotone_types: ");
```
Comment (StrikerRUS) [inline src/io/dataset.cpp]: `"\nmonotone_constraints: "`

File: src/io/dataset.cpp [hunk 6]
```diff
@@ -881,4 +941,71 @@ void Dataset::FixHistogram(int feature_idx, double sum_gradient, double sum_hess
...
+    for(size_t i = 0; i < dest_len; i++){
+      dest.push_back(deflt);
+    }
+    PushVector(dest, src);
+  }
+}
+
+void Dataset::addFeaturesFrom(Dataset* other){
+  if(other->num_data_ != num_data_){
+    throw std::runtime_error("Cannot add features from other dataset with a different number of rows");
```
Comment (StrikerRUS) [inline src/io/dataset.cpp]: Capitalize `Dataset`

File: src/io/dataset.cpp [hunk 7]
```diff
@@ -881,4 +941,71 @@ void Dataset::FixHistogram(int feature_idx, double sum_gradient, double sum_hess
...
+  PushVector(group_feature_cnt_, other->group_feature_cnt_);
+  feature_groups_.reserve(other->feature_groups_.size());
+  for(auto& fg : other->feature_groups_){
+    feature_groups_.emplace_back(new FeatureGroup(*fg));
+  }
+  for(auto feature_idx : other->used_feature_map_){
+    if(feature_idx >= 0){
+      used_feature_map_.push_back(feature_idx + num_features_);
+    } else {
+      used_feature_map_.push_back(-1); //Unused feature.
```
Comment (StrikerRUS) [inline src/io/dataset.cpp]: Two whitespaces before and one after `//`.

File: src/io/dataset.cpp [hunk 8]
```diff
@@ -881,4 +941,71 @@ void Dataset::FixHistogram(int feature_idx, double sum_gradient, double sum_hess
...
+    }
+  }
+  PushOffset(real_feature_idx_, other->real_feature_idx_, num_total_features_);
+  PushOffset(feature2group_, other->feature2group_, num_groups_);
+  auto bin_offset = group_bin_boundaries_.back();
+  //Skip the leading 0 when copying group_bin_boundaries.
```
Comment (StrikerRUS) [inline src/io/dataset.cpp]: One whitespace after `//`.

File: include/LightGBM/c_api.h [hunk 9]
```diff
@@ -269,14 +270,23 @@ LIGHTGBM_C_EXPORT int LGBM_DatasetGetFeatureNames(
...
                                              const char* filename);
 
+/*!
+* \brief save dataset to text file, intended for debugging use only
+* \param handle a instance of dataset
+* \param filename file name
+* \return 0 when succeed, -1 when failure happens
+*/
+LIGHTGBM_C_EXPORT int LGBM_DatasetDumpText(DatasetHandle handle,
+					   const char* filename);
```
Comment (StrikerRUS) [inline include/LightGBM/c_api.h]: Indent

File: include/LightGBM/c_api.h [hunk 10]
```diff
@@ -335,6 +345,15 @@ LIGHTGBM_C_EXPORT int LGBM_DatasetGetNumData(DatasetHandle handle,
...
                                                 int* out);
 
+/*!
+* \brief Add features from source to target, then free source
+* \param target The handle of the dataset to add features to
+* \param source The handle of the dataset to take features from
+* \return 0 when succeed, -1 when failure happens
+*/
+LIGHTGBM_C_EXPORT int LGBM_DatasetAddFeaturesFrom(DatasetHandle target,
+						  DatasetHandle source);
```
Comment (StrikerRUS) [inline include/LightGBM/c_api.h]: Indent

File: src/c_api.cpp [hunk 11]
```diff
@@ -813,6 +813,14 @@ int LGBM_DatasetSaveBinary(DatasetHandle handle,
   API_END();
 }
 
+int LGBM_DatasetDumpText(DatasetHandle handle,
+                           const char* filename) {
```
Comment (StrikerRUS) [inline src/c_api.cpp]: Indent

File: src/c_api.cpp [hunk 12]
```diff
@@ -878,6 +889,15 @@ int LGBM_DatasetGetNumFeature(DatasetHandle handle,
   API_END();
 }
 
+int LGBM_DatasetAddFeaturesFrom(DatasetHandle target,
+				 DatasetHandle source) {
```
Comment (StrikerRUS) [inline src/c_api.cpp]: Indent

File: python-package/lightgbm/basic.py [hunk 13]
```diff
@@ -1494,6 +1533,42 @@ def get_ref_chain(self, ref_limit=100):
                 break
         return ref_chain
 
+    def add_features_from(self, other):
+        """Add features from other to self.
+
+        This will add every feature from other to self and free other.
```
[the remark this answers is not in the document]
Comment (remcob-gr) [reply python-package/lightgbm/basic.py]: Fixed. The docstring was also outdated, this no longer frees other (see review comments above from @guolinke ).

**Comment by remcob-gr** [comment]
@StrikerRUS , thanks for the review, the issues should be fixed now.
Are there any others to take care of?

**Comment by remcob-gr** [comment]
Sorry for missing a few earlier, those should be all fixed now.

File: python-package/lightgbm/basic.py [hunk 14]
```diff
@@ -1494,6 +1533,47 @@ def get_ref_chain(self, ref_limit=100):
...
+        Parameters
+        ----------
+        filename : string
+            Name of the output file.
+
+        Returns
+        -------
+        self : Dataset
+            Returns self.
+
```
Comment (StrikerRUS) [inline python-package/lightgbm/basic.py]: Remove blank line.

**Comment by StrikerRUS** [comment]
@guolinke Can we merge?

**Comment by guolinke** [comment]
@StrikerRUS yes

**Pull request merged**